mysql-test/suite/innodb/t/lock_isolation.test

```diff
@@ -103,8 +102,46 @@ COMMIT;
 --connection con_weird
 --reap
 COMMIT;
---disconnect con_weird
 
 --connection default
 SELECT * FROM t;
 DROP TABLE t;
+
+--echo #
+--echo # MDEV-33802 Weird read view after ROLLBACK of other transactions
+--echo #
+
+CREATE TABLE t(a INT PRIMARY KEY, b INT UNIQUE) ENGINE=InnoDB;
+INSERT INTO t SET a=1;
+
+BEGIN; INSERT INTO t SET a=2;
+
+--connection consistent
+START TRANSACTION WITH CONSISTENT SNAPSHOT;
+--error ER_CHECKREAD
+SELECT * FROM t FORCE INDEX (b) FOR UPDATE;
```

Comment:
```
CURRENT_TEST: innodb.lock_isolation
mysqltest: At line 122: query 'SELECT * FROM t FORCE INDEX (b) FOR UPDATE' failed with wrong errno ER_LOCK_WAIT_TIMEOUT (1205): 'Lock wait timeout exceeded; try restarting transaction', instead of ER_CHECKREAD (1020)...
```
https://buildbot.mariadb.org/#/builders/534/builds/16014/steps/7/logs/stdio

Reply:
Even after applying some work-arounds (including `SET innodb_snapshot_isolation=ON` in `connection default`) I also reproduced the following bug in a test that precedes the one for this one:
```diff
CURRENT_TEST: innodb.lock_isolation
--- /mariadb/10.6/mysql-test/suite/innodb/r/lock_isolation.result	2024-04-09 10:28:08.988756438 +0300
+++ /mariadb/10.6/mysql-test/suite/innodb/r/lock_isolation.reject	2024-04-09 10:37:27.459725141 +0300
@@ -102,7 +102,7 @@
 connection default;
 SELECT * FROM t;
 a	b
-10	1
+10	20
 10	20
 DROP TABLE t;
 #

Result length mismatch
```
This is executed in a new transaction on its own, right after the execution of `COMMIT` in `con_weird`. Apparently, the `UPDATE` in `con_weird` can occasionally persistently corrupt the table. This is immediately before the results for this regression test.

After I moved the tests for this bug into a separate file, it passed 120,000 rounds of testing. I was so far unable to reproduce the failure that occurred on the CI system.

Reply:
`START TRANSACTION WITH CONSISTENT SNAPSHOT` does create a new read view in `innobase_start_trx_and_assign_read_view()` just fine. The bug is that the error is being ignored, and a new transaction would be implicitly started when using `./mtr --ps-protocol`. The following works around that problem.
```diff
diff --git a/mysql-test/suite/innodb/t/lock_isolation.test b/mysql-test/suite/innodb/t/lock_isolation.test
index 4c17afde7af..04a80fcae78 100644
--- a/mysql-test/suite/innodb/t/lock_isolation.test
+++ b/mysql-test/suite/innodb/t/lock_isolation.test
@@ -118,8 +118,10 @@ BEGIN; INSERT INTO t SET a=2;
 
 --connection consistent
 START TRANSACTION WITH CONSISTENT SNAPSHOT;
+--disable_ps_protocol
 --error ER_CHECKREAD
 SELECT * FROM t FORCE INDEX (b) FOR UPDATE;
+--enable_ps_protocol
 
 --connection con_weird
 START TRANSACTION WITH CONSISTENT SNAPSHOT;
```

Reply:
`--disable_ps2_protocol` turns out to be a sufficient work-around of [MDEV-33864](https://jira.mariadb.org/browse/MDEV-33864), which I just filed for this.